File: java/Puzzle.java
import java.util.List;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Collections;

public class Puzzle {

    /**
     * a b c
     * d e f
     * g h i
     * j k *
     */

    private static final int ROWS = 4, COLS = 3;

    private static List<String> getMoves(String subject)
    {
        List<String> moves = new LinkedList<>();

        int pos = subject.indexOf('*');
        int row = pos / COLS;
        int col = pos % COLS;

        // north
        if (row != 0) moves.add(swapChar(subject, pos, pos - COLS));
        // east
        if (col != COLS - 1) moves.add(swapChar(subject, pos, pos + 1));
        // south
        if (row != ROWS - 1) moves.add(swapChar(subject, pos, pos + COLS));
        // west
        if (col != 0) moves.add(swapChar(subject, pos, pos - 1));

        return moves;
    }

    private static String swapChar(String subject, int i, int j)
    {
        StringBuilder sb = new StringBuilder(subject);
        char tmp = sb.charAt(i);
        sb.setCharAt(i, sb.charAt(j));
        sb.setCharAt(j, tmp);
        return sb.toString();
    }

    public static List<String> solve(String puzzle, String goal)
    {
        for (int depth = 1; true; depth++) {
            LinkedList<String> root = new LinkedList<String>();
            root.add(puzzle);
            List<String> route = depthFirst(root, goal, depth);
            if (route != null) return route;
        }
    }

    private static List<String> depthFirst(LinkedList<String> route, String goal, int depth)
    {
        if (depth == 0) return null;

        String last = route.getLast();

        if (last.equals(goal)) return route;

        for (String move : getMoves(last)) {
            if ( ! route.contains(move)) {
                LinkedList<String> nextRoute = (LinkedList<String>) route.clone();
                nextRoute.add(move);
                List<String> fullRoute = depthFirst(nextRoute, goal, depth - 1);
                if (fullRoute != null) return fullRoute;
            }
        }

        return null;
    }

    private static String[] createPuzzle(int steps)
    {
        String puzzle = "", goal = "abcdefghijk*";

        for (int step = 0; step < steps; step++) {
            List<String> moves = getMoves(goal);
            Collections.shuffle(moves);
            if (step == steps / 2) puzzle = moves.get(0);
            goal = moves.get(0);
        }

        return new String[] { puzzle, goal };
    }

    public static void main(String[] args)
    {
        ArrayList<String[]> puzzles = new ArrayList<String[]>(10);
        for (int i = 0; i < 10; i++) {
            puzzles.add(createPuzzle(50));
        }

        for (String[] puzzle : puzzles) {
            List<String> solution = solve(puzzle[0], puzzle[1]);
            System.out.println(solution.size());
        }
    }

}

File: python/puzzle.py
import itertools

def id_dfs(puzzle, goal, moves):
    def dfs(route, depth):
        if depth == 0:
            return
        if goal(route[-1]):
            return route
        for move in moves(route[-1]):
            if move not in route:
                next_route = dfs(route + [move], depth - 1)
                if next_route:
                    return next_route

    for depth in itertools.count(1):
        route = dfs([puzzle], depth)
        if route:
            return route

def num_matrix_goal(rows, cols):
    '''Checks for the goal matrix based on supplied row and col length.'''
    def matrix(subject):
        for row in range(rows):
            for col in range(cols):
                value = subject[row][col]
                if value != 0 and value != row*rows + col+1:
                    return False
        return True
    return matrix

def num_matrix_moves(subject):
    moves = []

    zrow, zcol = next((r, c)
        for r, l in enumerate(subject)
            for c, v in enumerate(l) if v == 0)

    def board(row, col):
        import copy
        b = copy.deepcopy(subject)
        b[zrow][zcol] = b[row][col]
        b[row][col] = 0
        return b

    # north
    if zrow != 0:
        moves.append(board(zrow - 1, zcol))
    # east
    if zcol < len(subject[0]) - 1:
        moves.append(board(zrow, zcol + 1))
    # south
    if zrow < len(subject) - 1:
        moves.append(board(zrow + 1, zcol))
    # west
    if zcol != 0:
        moves.append(board(zrow, zcol - 1))

    return moves

def num_matrix(puzzle, steps):
    import random
    for step in range(steps):
        puzzle = random.choice(num_matrix_moves(puzzle))
    return puzzle

puzzle = num_matrix([[1, 2, 3], [4, 5, 6], [7, 8, 0]], 25)

solution = id_dfs(puzzle, num_matrix_goal(3, 3), num_matrix_moves)

print(len(solution))